libraries/chain_kv/include/b1/session/cache.hpp

```diff
@@ -0,0 +1,454 @@
+#ifndef cache_data_store_h
+#define cache_data_store_h
+
+#include <map>
+#include <unordered_set>
+#include <vector>
+
+#include <b1/session/bytes.hpp>
+#include <b1/session/cache_fwd_decl.hpp>
+#include <b1/session/key_value.hpp>
+
+namespace b1::session
+{
+
+// An in memory caching data store for storing key_value types.
+//
+// \tparam allocator The memory allocator to managed the memory used by the key_value instances stored in the cache.
+// \remarks This type implements the "data store" concept.
+template <typename allocator>
+class cache final
+{
+private:
+    // Currently there are some constraints that require the use of a std::map.
+    // Namely, we need to be able to iterator over the entries in lexigraphical ordering on the keys.
+    using cache_type = std::map<bytes, key_value>;
+
+public:
+    using allocator_type = allocator;
+    
+    class iterator;
+    using const_iterator = const iterator;
+    
+    class iterator  final: public std::iterator<std::bidirectional_iterator_tag, key_value>
```

Comment:
std::iterator is deprecated in C++17

Reply:
https://www.fluentcpp.com/2018/05/08/std-iterator-deprecated/